rpcclient/rawtransactions.go

```diff
@@ -880,3 +881,134 @@ func (c *Client) DecodeScriptAsync(serializedScript []byte) FutureDecodeScriptRe
 func (c *Client) DecodeScript(serializedScript []byte) (*btcjson.DecodeScriptResult, error) {
 	return c.DecodeScriptAsync(serializedScript).Receive()
 }
+
+// FutureTestMempoolAcceptResult is a future promise to deliver the result
+// of a TestMempoolAccept RPC invocation (or an applicable error).
+type FutureTestMempoolAcceptResult chan *Response
+
+// Receive waits for the Response promised by the future and returns the
+// response from TestMempoolAccept.
+func (r FutureTestMempoolAcceptResult) Receive() (
+	[]*btcjson.TestMempoolAcceptResult, error) {
+
+	response, err := ReceiveFuture(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal as an array of TestMempoolAcceptResult items.
+	var results []*btcjson.TestMempoolAcceptResult
+
+	err = json.Unmarshal(response, &results)
+	if err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+// TestMempoolAcceptAsync returns an instance of a type that can be used to get
+// the result of the RPC at some future time by invoking the Receive function
+// on the returned instance.
+//
+// See TestMempoolAccept for the blocking version and more details.
+func (c *Client) TestMempoolAcceptAsync(txns []*wire.MsgTx,
+	maxFeeRate float64) FutureTestMempoolAcceptResult {
+
+	// Due to differences in the testmempoolaccept API for different
+	// backends, we'll need to inspect our version and construct the
+	// appropriate request.
+	version, err := c.BackendVersion()
+	if err != nil {
+		return newFutureError(err)
+	}
+
+	log.Debugf("TestMempoolAcceptAsync: backend version %s", version)
+
+	// Exit early if the version is below 22.0.0.
+	//
+	// Based on the history of `testmempoolaccept` in bitcoind,
+	// - introduced in 0.17.0
+	// - unchanged in 0.18.0
+	// - allowhighfees(bool) param is changed to maxfeerate(numeric) in
+	//   0.19.0
+	// - unchanged in 0.20.0
+	// - added fees and vsize fields in its response in 0.21.0
+	// - allow more than one txes in param rawtx and added package-error
+	//   and wtxid fields in its response in 0.22.0
+	// - unchanged in 0.23.0
+	// - unchanged in 0.24.0
+	// - added effective-feerate and effective-includes fields in its
+	//   response in 0.25.0
+	//
+	// We decide to not support this call for versions below 22.0.0. as the
+	// request/response formats are very different.
+	if version < BitcoindPre22 {
+		err := fmt.Errorf("%w: %v", ErrBitcoindVersion, version)
+		return newFutureError(err)
+	}
+
+	// The maximum number of transactions allowed is 25.
+	if len(txns) > 25 {
+		err := fmt.Errorf("%w: too many transactions provided",
+			ErrInvalidParam)
+		return newFutureError(err)
+	}
+
+	// Exit early if an empty array of transactions is provided.
+	if len(txns) == 0 {
+		err := fmt.Errorf("%w: no transactions provided",
```

Comment:
do we need to do these checks? ie, wont bitcoind do these checks for us anyways?

Reply:
Commented below - mostly due to we need to also maintain the same behavior for `btcd`.

Reply:
gotcha! 👍